api.go

```diff
@@ -0,0 +1,469 @@
+	// Beta is the beta version of the model to use for the sample.
+	Beta string `json:"beta"`
+	// MaxTokens is the mas number of tokens to generate.
+	MaxTokens int `json:"max_tokens"`
+	// The temperature to use for sampling.
+	Temperature float64 `json:"temperature"`
+}
+
+// A Conversation is a sequence of messages between a user and an assistant.
+type Conversation struct {
+	// System is the system prompt to use for the conversation.
+	System *string
+	// Messages is the sequence of messages in the conversation. This always
+	// starts with a user message, and alternates between user and assistant.
+	Messages *[]*Message
+	// Usage is the usage statistics for the conversation in total.
+	Usage struct {
+		InputTokens  int `json:"input_tokens"`
+		OutputTokens int `json:"output_tokens"`
+	}
+	// HttpClient is the HTTP client used for API requests
+	HttpClient *http.Client
+	// apiToken is the API token used for API requests
+	ApiToken string `json:"-"`
+	// Settings is the settings for the conversation.
+	Settings *SampleSettings
+}
+
+// NewConversation creates a new conversation with the given system prompt. It
+// initializes the messages and usage statistics, as well as reasonable
+// defaults.
+func NewConversation(system string) *Conversation {
+	messages := make([]*Message, 0)
+	conversation := Conversation{
+		System:   &system,
+		Messages: &messages,
+	}
+	conversation.Usage.OutputTokens = 0
+	conversation.Usage.InputTokens = 0
+	// Copy our default settings
+	settings := DefaultSettings
+	conversation.Settings = &settings
+	conversation.ApiToken = DefaultApiToken
+
+	// Initialize the HTTP client
+	conversation.HttpClient = &http.Client{}
+	return &conversation
+}
+
+// AddMessage adds a message to the conversation with the given role and
+// content. It used internally, and also can be used externally to
+// manipulate conversations.
+func (conversation *Conversation) AddMessage(
+	role string,
+	content *[]ContentBlock,
+) {
+	message := Message{
+		Role:    role,
+		Content: content,
+	}
+	*conversation.Messages = append(*conversation.Messages, &message)
+}
+
+// Messages API URI
+// POST /v1/messages
+var messagesURI = "https://api.anthropic.com/v1/messages"
+
+// API default headers
+var headers = map[string]string{
+	"Content-Type":      "application/json",
+	"x-api-key":         "",
+	"anthropic_version": "2023-06-01",
+	// "anthropic-beta":    "max-tokens-3-5-sonnet-2024-07-15",
+}
+
+var retries = 3
+var retryDelay = 3 * time.Second
+
+// Send sends a message to the assistant and returns the reply. It also
+// returns the reason the conversation stopped, the number of input tokens
+// used, and the number of output tokens used.
+//
+// If the text is empty, it sends the message as is, and does not add a user
+// message to the conversation. This is useful for continuing an incomplete
+// conversation by "assistant", in the case of a stopReason of "max_tokens".
+func (conversation *Conversation) Send(text string) (
```

Comment:
API design suggestion: why not return a struct, so that possible future fields can be easily extended in a non-breaking manner? (applies to `SendUntilDone` as well)

Reply:
🤔 What sort of struct? It updates the `Conversation` itself, as it has state that needs to be retained between submissions.

Do you mean returning a new `Conversation`, keeping the original one immutable, or perhaps returning a `Response`?

Reply:
I mean the latter, returning `Response` itself, rather than an unpacked 5-tuple version

Reply:
That makes sense, as that means more metadata can be returned that way. Thank you for the idea.